### PR TITLE
libpeas: add support for running python plugins

### PR DESCRIPTION
**Pull request by c-alpha**
Added a new option "--with-python" which builds libpeas with the
python runtime to run python 2 and 3 plugins.

This PR is e.g. required in order to be able to build [astroidmail/astroid](https://github.com/astroidmail/astroid).

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by c-alpha** [comment]
@ilovezfs ?

**Comment by ilovezfs** [comment]
@DomT4 can you help with backtick cleansing

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: You've made it an option here but the default below. Can only be one or the other 😄.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: `depends_on "python"` and `depends_on "python3"`. The syntax here is wrong.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: Use `Utils.popen_read()` rather than backticks.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: `ENV.append "LDFLAGS", "-L#{py2_lib} -L#{py3_lib}"`.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: It won't stretch past 80 character so just:
```ruby
args << "--enable-python2" << "--enable-python3"
```

Comment (DomT4) [reply Formula/libpeas.rb]: Also if this becomes a complete default you can inline this rather than creating an `args` array, but that's up to @ilovezfs.

**Review by DomT4 — CHANGES_REQUESTED** [review]

File: Formula/libpeas.rb
Comment (ilovezfs) [inline Formula/libpeas.rb]: I think this isn't set anymore, is it?

Comment (DomT4) [reply Formula/libpeas.rb]: IIRC as long as Homebrew's `python` is used rather than the system's.

Comment (ilovezfs) [reply Formula/libpeas.rb]: I mean … there is no python req anymore so … where would it be set?

Comment (DomT4) [reply Formula/libpeas.rb]: _Potentially_ when building from source, there's a `with/without-python` option and Homebrew's `python` isn't installed, if I'm reading [this blob](https://github.com/Homebrew/brew/blob/5496a04dd977b1e0d6c71ed0ff6de01e722ca513/Library/Homebrew/language/python.rb#L16-L29) correctly. But, 🤷‍♂️.

Comment (DomT4) [reply Formula/libpeas.rb]: I get what you're saying though. Practically speaking it should never be set by default any more.

Comment (ilovezfs) [reply Formula/libpeas.rb]: Ug. That's depressing

**Comment by c-alpha** [comment]
First of all, many thanks for your swift response and kind support in improving the PR, @ilovezfs and @DomT4! This is both, my first ruby, and my first formula. So many thanks for bearing with me, too.

I _believe_ to have addressed all of @DomT4's improvement suggestions with b73bde4f8936400cf989546210a352d28bdcc684. Happy to stand corrected, of course.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: Not sure it wouldn't follow Homebrew's convention better to separate out the `python`s here into `python` and `python3` options. Throwing to @ilovezfs for an opinion on that.

Comment (c-alpha) [reply Formula/libpeas.rb]: `libpeas` can support both, python 2 and 3 in parallel. I would also assume that most people will have both python versions installed.

**Comment by c-alpha** [comment]
@DomT4, are you happy with the changes I made with b73bde4f8936400cf989546210a352d28bdcc684, and which I believe should address your initial comments?

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: That was to say, I believe a single switch is all that's needed.

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: The unsetting was certainly needed on my system (else I wouldn't have put it there in the first place).

If it's no longer set on "modern" installs, then the unsetting wouldn't harm either, while still allowing it wo work on "classic" installs?

Comment (ilovezfs) [reply Formula/libpeas.rb]: @c-alpha there's no such thing as "modern" vs. "classic" as we only support the HEAD of homebrew/core.

File: Formula/libpeas.rb
Comment (ilovezfs) [reply Formula/libpeas.rb]: Yeah I'd separate it out. `--with-python` implies python2 specifically in Homebrew currently.

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: Sorry for having been unclear. I was referring to the python install. If you are using the version that ships with macOS or still have a manually installed one somewhere from yesteryear ("classic"), then you might need the `ENV.unset` bit. If you have python installed from homebrew ("modern"), then you may not need it.

So in summary, it would seem that by dropping the `ENV.unset` line, we could potentially break the install for a certain clientele?

Comment (ilovezfs) [reply Formula/libpeas.rb]: we filter out all your environment variables

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: I did some experiments, trying to separate `python2` and `python3` into separate switches.

There are two PyGObject formulae: `pygobject` and `pygobject3`.
* `brew install pygobject` installs [version 2.28.6 from June 2011](https://pygobject.readthedocs.io/en/latest/changelog.html#id102)
* `brew install pygobject3` installs [version 3.26.1 from October 2017](https://pygobject.readthedocs.io/en/latest/changelog.html#id4) with support for `python2` only

When installing `libpeas --with-python`, and with a dependency on `pygobject`, the `configure` script of `libpeas` fails, complaining that PyGObject was not installed. So it seems the current `libpeas` no longer works with the 2011 `pygobject` version.

When installing `libpeas --with-python` only, and with a dependency on `pygobject3`, it depends on whether `pygobject3` was built `--with-python3`. (NOTE: `python2` support is always built into `pygobject3`, unless excluded with `--without-python`. END NOTE)

`libpeas` | `pygobject3` | `pygobject3 --with-python3`
------- | ------------ | ------------------------------
`--with-python` | :one: builds ok | :two: build fails 
`--with-python3` | :three: build fails  | :four: builds ok 
`--with-python --with-python3` | :five: build fails | :six: builds ok 

So there are three user stories we have to consider:
1. The user has installed `python2` only, and therefore has installed `pygobject3` without `python3` support. In this case, `libpeas --with-python` builds fine.
1. The user has installed `python3` only, and therefore has installed `pygobject3` with `python3` support, but without `python2` support. In this case, `libpeas --with-python3` builds fine.
1. The user has installed `python2` and `python3` in parallel, and therefore has installed `pygobject3` with support for both, `python2` and `python3`. In this case, all combinations will work, **except** when he tries to compile `libpeas --with-python`, i.e. with `python2` support only, despite having `python3` already installed. This corresponds to :two: in the table above.

I don't know what a motivation could be for doing user story no. 3, but then, you never know.

What could possible approaches be for making :two: work, short of fixing `pygobject3`? If we specify a `libpeas` build-only dependency on `pygobject3` sans `python3` support, it would build, but presumably fail at runtime.

The two easiest options would seem:
* leave as is now, i.e. always build support for both python versions;
* or accept :two: as a corner case that can be ignored since people doing this don't deserve any better anyway.

Looking forward to your thoughts.

**Comment by c-alpha** [comment]
@ilovezfs ? @DomT4 ? Any more thoughts of yours? Can this be merged?

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: Can this be built against the system `python`? If it can we could just turn it on by default, and leave the `python3` part as an option.

File: Formula/libpeas.rb
Comment (DomT4) [inline Formula/libpeas.rb]: We know this variable since we're always using Homebrew's `python3` now. It's:
```ruby
Formula["python3"].opt_frameworks/"Python.framework/Versions/#{pyver}/lib
```

So let's just do:
```ruby
if build.with? "python3"
  py3_lib = Formula["python3"].opt_frameworks/"Python.framework/Versions/#{pyver}/lib
  ENV.append "LDFLAGS", "-L#{py3_lib}"
  args << "--enable-python3"
end
```

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: My take would be to allow people to only install the languages they want/need. And the previous default was to build without any language support. In that light, I would suggest to keep the previous default behaviour.

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: Done with d7408de76d47c530ae70e4b51d299922607150a3. Some extra tweaking was necessary, however, to get the version part of the library path right.

**Comment by c-alpha** [comment]
@ilovezfs ? @DomT4 ? Any more thoughts of yours? Can this be merged?

**Comment by c-alpha** [comment]
@ilovezfs ? @DomT4 ?

File: Formula/libpeas.rb
Comment (ilovezfs) [inline Formula/libpeas.rb]: I think if we have explicit option text up here for `"with-python"` then we should also have explicit option text up here for `"with-python3"`

File: Formula/libpeas.rb
Comment (ilovezfs) [reply Formula/libpeas.rb]: @c-alpha it doesn't add additional dependencies if we use system Python so we should just do that and make python3 optional.

```
depends_on "python" if MacOS.version <= :snow_leopard
```

Then always  build the python 2 bindings and build the python3 bindings if --with-python3 is passed too.

File: Formula/libpeas.rb
Comment (ilovezfs) [reply Formula/libpeas.rb]: Actually there will be no `"with-python"` option so we just need option text for python3.

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: Implemented with cb02cfdf69c2b4a501f3614463c3844ffff88925.

File: Formula/libpeas.rb
Comment (c-alpha) [reply Formula/libpeas.rb]: Implemented with cb02cfdf69c2b4a501f3614463c3844ffff88925.

**Comment by c-alpha** [comment]
@ilovezfs ? @DomT4 ?

**Comment by quantumproducer** [comment]
Bump

**Comment by c-alpha** [comment]
Thanks for the bump, @quantumproducer. I have been through a couple of iterations with the maintainers already.

From my side, this would be ready for merging.

@ilovezfs, @DomT4, any further thoughts? Could this be merged?

**Comment by tschoonj** [comment]
This needs a bit more work now that `pygobject3` uses `python` as `:recommended`, and `python@2` became `:optional`

**Comment by c-alpha** [comment]
> This needs a bit more work now that `pygobject3` uses `python` as `:recommended`, and `python@2` became `:optional`

Thanks for the heads-up, @tschoonj.

So I take you suggest having similar semantics for `libpeas`, too? Currently, `python@2` is required as preferred by @DomT4. So my immediate impulse would be to simply bump `python3` from `:optional` to `:recommended`.  Would that cover your point?

**Comment by tschoonj** [comment]
Well `python3` is now just `python`, as it became the preferred python interpreter for Homebrew recently.

And yes, it would be good to use a similar options scheme as `pygobject3`, which means that `python@2` would be come `:optional`

**Comment by c-alpha** [comment]
> And yes, it would be good to use a similar options scheme as `pygobject3`, which means that `python@2` would be come `:optional`

I see. This would be fine by me.

@DomT4, you had advocated for building with `python@2` by default; any views of yours?

**Comment by c-alpha** [comment]
@tschoonj, @DomT4, have you had any chance to look into agreeing a preferred way forward?

I would be fine with either approach, i.e. either have `python@2` by default, and an extra switch for `python@3` (as preferred by @DomT4), or to have `python@3` by default, and an extra switch for `python@2` (as preferred by @tschoonj).

@ilovezfs, any views of yours?

**Comment by ilovezfs** [comment]
@c-alpha since pygobject default is now python 3, I'd say python 3 default here, too.

**Comment by c-alpha** [comment]
@ilovezfs wrote:
> @c-alpha since pygobject default is now python 3, I'd say python 3 default here, too.

Done. And thanks for sharing your helpful thoughts!

**Comment by c-alpha** [comment]
bump...

**Comment by c-alpha** [comment]
bump...
_[I guess @quantumproducer would support my "bump"? ;-)]_

There haven't been any comments from the usual suspects (@ilovezfs, @DomT4, @tschoonj), or otherwise, in quite a while since my last commits.

As mentioned in my description of this PR:

> This PR is e.g. required in order to be able to build [astroidmail/astroid](https://github.com/astroidmail/astroid).

So it would be **really much appreciated**, if we could move forward with this PR now, **and seriously consider merging it, please.**

**Comment by ilovezfs** [comment]
Which bindings does astroidmail/astroid need?

**Comment by c-alpha** [comment]
> Which bindings does astroidmail/astroid need?

`python@2` and `python@3`.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Going to waive the audit
```
  * C: 26: col 33: Dependency pygobject3 should not use option with-python
```
because we can fix it whenever we fix `mat` too.

**Comment by ilovezfs** [comment]
Thanks so much for your first contribution (and patience)! Without people like you submitting PRs we couldn't run this project. You rock, @c-alpha!

**Comment by c-alpha** [comment]
> Thanks so much for your first contribution (and patience)!

Thanks for your patience, too!

> Without people like you submitting PRs we couldn't run this project. You rock, @c-alpha!

Glad you found my contribution useful.

Many thanks to the reviewers @ilovezfs and @DomT4, and also to the other commenters (@quantumproducer, @tschoonj)! Looking forward to my next PR.

**Comment by DomT4** [comment]
Thanks @c-alpha; the patience working on this through completion is appreciated.

**Pull request closed without merging**
